### PR TITLE
Interaction

### PR DESCRIPTION
**Pull request by houpc**
add the calculation of 2D polarization and self-energy in zero temperature.

### CODE REVIEW

File: src/polarization.jl [hunk 1]
```diff
@@ -8,30 +8,63 @@ using ..Convention
...
+#     end
+
+#     return density * integrand
+# end
+
+# Analytical calculated integrand of Π0 in 2D.
+@inline function _ΠT2d_integrand(k, q, ω, param)
+    @unpack me, β, EF = param
+    density = me / 2π
+    nk = 1.0 / (exp(β * (k^2 / 2 / me - EF)) + 1)
```
Comment (kunyuan) [inline src/polarization.jl]: EF -> \mu ? For finite temperature, \mu and EF are not the same!

File: src/polarization.jl [hunk 2]
```diff
@@ -8,30 +8,63 @@ using ..Convention
...
+
+    return nothing
+end
+
+# Analytical calculated integrand of Π0 in 3D.
+@inline function _ΠT3d_integrand(k, q, ω, param)
+    @unpack me, β, EF = param
     # ω only appears as ω^2 so no need to check sign of ω
+    nk = 1.0 / (exp(β * (k^2 / 2 / me - EF)) + 1)
```
Comment (kunyuan) [inline src/polarization.jl]: EF -> \mu ? For finite temperature, \mu and EF are not the same!

File: src/polarization.jl [hunk 3]
```diff
@@ -8,30 +8,63 @@ using ..Convention
...
+@inline function _ΠT2d_integrand(k, q, ω, param)
+    @unpack me, β, EF = param
+    density = me / 2π
+    nk = 1.0 / (exp(β * (k^2 / 2 / me - EF)) + 1)
+
+    return nothing
+end
+
+# Analytical calculated integrand of Π0 in 3D.
+@inline function _ΠT3d_integrand(k, q, ω, param)
```
Comment (kunyuan) [inline src/polarization.jl:52]: \omega is an integer or float? If it is an integer, explicitly specify the type.

File: src/polarization.jl [hunk 4]
```diff
@@ -46,7 +79,7 @@ end
 """
     function Polarization0_FiniteTemp(q, n, param, maxk=20, scaleN=20, minterval=1e-6, gaussN=10)
 
-Finite temperature Π0 function for matsubara frequency and momentum. Analytically sum over transfer frequency and angular
+Finite temperature one-spin Π0 function for matsubara frequency and momentum. Analytically sum over transfer frequency and angular
 dependence of momentum, and numerically calculate integration of magnitude of momentum.
 Slower(~200μs) than Polarization0_ZeroTemp.
 Assume G_0^{-1} = iω_n - (k^2/(2m) - E_F)
```
Comment (kunyuan) [inline src/polarization.jl]: EF -> \mu

File: src/polarization.jl [hunk 5]
```diff
@@ -8,30 +8,63 @@ using ..Convention
...
+# Analytical calculated integrand of Π0 in 3D.
+@inline function _ΠT3d_integrand(k, q, ω, param)
+    @unpack me, β, EF = param
     # ω only appears as ω^2 so no need to check sign of ω
+    nk = 1.0 / (exp(β * (k^2 / 2 / me - EF)) + 1)
 
     # if q is too small, use safe form
-    if q < 1e-16 && ω==0
-        if abs(q-2*k)^2<1e-16
+    if q < 1e-16 && ω == 0
```
Comment (kunyuan) [inline src/polarization.jl]: \omega is a float, use "\omega \approx 0" to compare a float

File: src/polarization.jl [hunk 6]
```diff
@@ -8,30 +8,63 @@ using ..Convention
...
+        if abs(q - 2 * k)^2 < 1e-16
             return 0.0
         else
-            return k*me/(2*π^2)/(exp(β*(k^2/2/me-EF))+1)*((8*k)/((q-2*k)^2))
+            return -k * me / (4 * π^2) * nk * ((8 * k) / ((q - 2 * k)^2))
         end
-    elseif q < 1e-16 && ω!=0
-        return k*me/(2*π^2)/(exp(β*(k^2/2/me-EF))+1)*((8*k*q^2)/(4*me^2*ω^2+(q^2-2*k*q)^2))
+    elseif q < 1e-16 && ω != 0
```
Comment (kunyuan) [inline src/polarization.jl]: (\omega \approx 0) == false

File: src/selfenergy.jl [hunk 7]
```diff
@@ -8,91 +8,140 @@ using ..Parameter, ..Convention, ..Polarization, ..Interaction, ..LegendreIntera
...
+    # container of Σ
+    Σ = GreenFunc.Green2DLR{ComplexF64}(:sigma, GreenFunc.IMTIME, β, true, fdlr.Euv, kgrid, 1)
+    Σ_ins = zeros(ComplexF64, (1, 1, length(kgrid.grid)))
+    Σ_dyn = zeros(ComplexF64, (1, 1, length(kgrid.grid), fdlr.size))
+
+    # equal-time green (instant)
+    G_ins = tau2tau(G.dlrGrid, G.dynamic, [β,], G.timeGrid.grid; axis = 4)[1, 1, :, 1]
+    @assert length(G.instant) == 0 "current implication supports green function without instant part"
+
+    for (ki, k) in enumerate(kgrid.grid)
```
Comment (kunyuan) [inline src/selfenergy.jl:67]: iterate ki first, then ti. Julia is column first language like fortran

File: src/polarization.jl [hunk 8]
```diff
@@ -60,84 +93,177 @@ Assume G_0^{-1} = iω_n - (k^2/(2m) - E_F)
...
+            @assert !isnan(integrand[ki]) "nan at k=$k, q=$q"
+        end
+    elseif dim == 3
+        mink = (q < 1e-16 / minterval) ? minterval * kF : minterval * min(q, kF)
+        kgrid = CompositeGrid.LogDensedGrid(:gauss, [0.0, maxk * kF], [0.5 * q, kF], scaleN, mink, gaussN)
+        integrand = zeros(Float64, kgrid.size)
+        for (ki, k) in enumerate(kgrid.grid)
+            integrand[ki] = _ΠT3d_integrand(k, q, 2π * n / β, param)
+            @assert !isnan(integrand[ki]) "nan at k=$k, q=$q"
+        end
```
Comment (iintSjds) [inline src/polarization.jl:95]: don't duplicate code. use
if dim == 2
    integrand_func=_ΠT2d_integrand
elseif dim == 3
    integrand_func=_ΠT3d_integrand
end
...
integrand[ki] = integrand_func(k, q, 2π * n / β, param)

**Pull request merged**